docs/rfcs/0003-initial.md

```diff
@@ -0,0 +1,280 @@
+`number` and `string`.
+
+## Database Consumer API (`x/database/sql/sql.ts`)
+
+Inspired by [`sql.go`](https://golang.org/src/database/sql/sql.go), this is
+primary interface, which most users will use. `Value` below represents the
+driver-defined `Value` type for bindings and results. The interface is entirely
+async for now, even if the underlying driver supports sync operations.
+
+- `sql.open(path, driver): Promise<sql.Database>`
+  - Creates a database handle/connector with the given driver and path. May
+    validate the arguments (path), but will not open a connection yet.
+- `sql.Database` class
+  - `.connect(): Promise<sql.Connection>`
+    - Opens a new open connection to the database. In the future a connection
+      pool will be maintained instead of always opening new connections.
+- `sql.Connection` class
+  - `.startTransaction(): Promise<sql.Transaction>`
+    - Starts a new transaction in the connection. If a top-level transaction is
+      already in progress on this connection, this will block until it is
+      finished.
+  - `.close(): Promise<void>`
+    - Closes the connection. If there is an active transaction, this will block
+      until it is finished.
+- `sql.Transaction` class
+  - `.prepareStatement(query: string): Promise<sql.PreparedStatement>`
+    - Prepares a SQL query for execution in this transaction.
+  - `.commit(): Promise<void>`
+    - Completes the transaction with changes committed.
+  - `.rollback(): Promise<void>`
+    - Completes the transaction with changes rolled back.
+  - `.startTransaction(): Promise<sql.Transaction>`
+    - Starts a nested transaction within this transaction. If a nested
+      transaction is already in progress, this will block until it is finished.
+      While a nested transaction is in progress, queries should be executed
+      through the inner-most active transaction, not the parent transaction.
+- `sql.PreparedStatement` class
+  - `.query(args?): AsyncGenerator<Iterator<Value>>`
+    - Executes the query with an optional array of bound values, and
+      incrementally reads rows from the database. The iterator should be
+      disposed of by calling `.return()` (which a `for await` statement will do
+      automatically).
+  - `.queryRow(args?): Promise<Array<Value>>`
+    - Executes the query with an optional array of bound values, returning only
+      the first row of results, as an array.
+  - `.exec(args?): Promise<{ insertedRowId?: Value, affectedRowCount?: number }>`
+    - Executes the query with an optional array of bound values, without
+      returning any result rows. A `insertedRowId` and `affectedRowCount` value
+      may be returned, but note that for some drivers these values may reflect a
+      previous query if the executed one did not actually insert or affect any
+      rows. (These should only be absent if the driver is certain that they're
+      not relevant to executed query, or doesn't support them at all.)
+  - `.dispose(): Promise<void>`
```

Comment:
What is the purpose of this method? I assume to close the prepared statement?

Reply:
Yes, something like this is needed to close the statement/ collect the associated resources on the DB server/ embedded database.

E.g. in the sqlite package, the `.query()` method returns a `Rows`, and the `Rows` will dispose the underlying statement, once the iterator returns.